### PR TITLE
Fix lazy loading of enum static initializers

### PR DESCRIPTION
**Pull request by antlechner**
Rather than calling clinit_wrapper on every parameter to the entry
method that is an enum, we should first load all the classes that may
be needed before we enter the entry method, and then call
clinit_wrapper on all of those which are enums. For example, the case of
an enum field of an argument to the entry method would have previously
hit an invariant violation.

### CODE REVIEW

File: jbmc/regression/jbmc/NondetEnumArgField/test.desc [hunk 1]
```diff
@@ -0,0 +1,9 @@
+CORE
+NondetEnumArgField.class
+--function NondetEnumArgField.test --cp `../../../../scripts/format_classpath.sh . ../../../src/java_bytecode/library/core-models.jar`
```
Comment (mgudemann) [inline jbmc/regression/jbmc/NondetEnumArgField/test.desc:3]: why can't the field be `null`?

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp [hunk 2]
```diff
@@ -407,28 +397,35 @@ void ci_lazy_methodst::initialize_instantiated_classes(
   // As in class_loader, ensure these classes stay available
   for(const auto &id : extra_instantiated_classes)
     needed_lazy_methods.add_needed_class("java::" + id2string(id));
+
+  // Special case for enums. We may want to generalise this, see TG-4689
+  // and the comment in java_object_factoryt::gen_nondet_pointer_init.
+  for(const auto &found_class : needed_lazy_methods.get_instantiated_classes())
+  {
+    const auto &class_type = to_java_class_type(ns.lookup(found_class).type);
+    if(class_type.get_base("java::java.lang.Enum"))
```
Comment (mgudemann) [inline jbmc/src/java_bytecode/ci_lazy_methods.cpp:406]: I think we can test on `ID_enumeration` for class types instead of this comparison.

**Review by smowton — APPROVED** [review]

**Review by allredj — APPROVED** [review]
Makes sense, even to me!

File: jbmc/src/java_bytecode/ci_lazy_methods.cpp (hunk 2 above)
Comment (antlechner) [reply jbmc/src/java_bytecode/ci_lazy_methods.cpp:406]: Huh, I've been doing this all over the place and no one told me about `ID_enumeration`. :confused: I'd like to get this merged quickly to fix my previous mistake, happy if I change all the `get_base`s in a follow-up PR?

File: jbmc/regression/jbmc/NondetEnumArgField/test.desc (hunk 1 above)
Comment (antlechner) [reply jbmc/regression/jbmc/NondetEnumArgField/test.desc:3]: It can, I just wanted it to assert that it can in fact come up with a non-null enum.

**Pull request merged**